### PR TITLE
Add a private _safe_softmax

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Changes the stance of SDPA on what to do for fully masked out rows

## Current Behavior
Several PyTorch users have expressed frustration over this issue:
- https://github.com/pytorch/pytorch/issues/41508
- https://github.com/pytorch/pytorch/issues/103749
- https://github.com/pytorch/pytorch/issues/103963

These are significant issues with extensive discussion but no satisfactory resolution. The PyTorch team's consensus, as stated here:
https://github.com/pytorch/pytorch/issues/24816#issuecomment-524415617

Can be paraphrased as follows:

When passing in fully masked out rows, attention becomes ambiguous. We have two main options:

1. Uniformly attend to all values:
   ```python
   scores[masked_out_rows] = 1 / len(row)
   out[masked_out_rows] = 1 / len(row) * value
   ```

2. Decide that attention between no queries (masked) and no keys (masked) is meaningless:
   ```python
   output[fully_masked_rows] = NaN
   ```

We went with option 2. Partially because it was easier to implement, but also people argued that users can slice the output to remove the NaNs:
``` Python
>fill_value = -float("inf") 
>row0 = torch.randn(4)
>row1 = torch.tensor([(fill_value for _ in range(4)])
>matrix = torch.stack([row0, row1]).requires_grad_(True)
>out = torch.softmax(matrix, 1)
>out = out[0]
>print(out)
tensor([0.5377, 0.2729, 0.0692, 0.1201])
```
Cool, problem solved. But what happends when you call backwards..
```Python
>out.backward(torch.ones_like(out))
>print(matrix.grad)
tensor([[3.0957e-08, 1.4157e-08, 7.7802e-10, 1.3713e-08],
        [       nan,        nan,        nan,        nan]])
```
Those pesky NaNs are back!

## Why do we see NaNs today?

The core of the problem revolves around using softmax function in sdpa:

```python
> row = torch.tensor([(-float("inf")) for _ in range(4)])
> torch.softmax(row, 0)
tensor([nan, nan, nan, nan])
```

## Quick Aside: Masking in Attention

Attention itself doesn't have a concept of masking. The `sdpa` function has an argument called `attn_mask`, which would be more accurately named `attn_bias`. This is because we don't actually "mask" entries when computing attention. Instead, due to implementation details([performance](https://github.com/pytorch/pytorch/issues/25110#issuecomment-524519087)), we add a value to the masked-out query/key pairs.

We use a large negative number (typically -inf) to decrease the attention weight, as softmax assigns more weight to larger values.

## Alternative Approaches

If we use a very large negative number instead of -inf:

```python
> row = torch.tensor([(-1e6) for _ in range(4)])
> torch.softmax(row, 0)
tensor([0.2500, 0.2500, 0.2500, 0.2500])
```
However if users always remembered to "slice" out their outputs i.e.:
```Python
>fill_value = -1e6
>...
>out.backward(torch.ones_like(out))
>print(matrix.grad)
tensor([[-0.0563, -0.0564,  0.1613, -0.0486],
        [ 0.0000,  0.0000,  0.0000,  0.0000]])
```
This would bring us back into a better state.

## A Third Option

We don't necessarily need to alter the behavior of softmax for -inf or very large negative numbers. The fundamental goal is to exclude certain query/key pairs from attention, regardless of the underlying implementation.

This PR implements the new semantic for masking w/ attention in fully masked-out rows:
```python
out[masked_out_rows] = 0
```

**Important Note**: This idea isn't entirely new. The [MaskedTensor](https://pytorch.org/tutorials/prototype/maskedtensor_overview#safe-softmax) prototype, a tensor subclass, was designed to handle such cases. However, it remains a prototype feature and hasn't gained widespread adoption.

## Details
This PR stack does 3 things:
1. Adds a PRIVATE _safe_softmax op
2. Updates semantic for flash_cpu fused kernel
3. Updates semantic for efficient_cuda fused kernel

_safe_softmax is not supposed to be used generically and is only meant to be used within the context of SDPA. Due to this fact instead of decomposing softmax and checking for -inf rows we instead "cheat" and use nan_to_num.

Why I think this is okay? (please find a counter point if avail)
There are multiple ways NaNs can emerge. For the fully masked out rows case nan_to_num works. But what if there were other NaNs, wouldn't this silently remove them? 

The only case that this can happen is if the input itself had a NaN or an Inf
For example: 
```Python
a = torch.ones([4], requires_grad=False, dtype=torch.float16)
a[1] = torch.finfo(torch.float16).max
print(a.softmax(-1))
```
Will return 
`tensor([0., 1., 0., 0.], dtype=torch.float16)`

Where 
```Python
a = torch.ones([4], requires_grad=False, dtype=torch.float16)
a[1] = float("inf")
a.softmax(-1)
```
returns:
`tensor([nan, nan, nan, nan], dtype=torch.float16)`

If we dont want to even allow for the possibility of "inf" or "NaN" attention scores to be converted to 0 then we can implemented it something like this

```Python
max = torch.max(a, dim=-1, keepdim=True)
exp = torch.exp(a - max.values)
denom = torch.sum(exp, dim=-1, keepdim=True)
softmax = exp / denom
softmax = torch.where(max.values == float('-inf'), 0.0, softmax)
```
however we would be paying for this in math performance.

## Why Now
I think one point that has substantially changed where PyTorch should lie on this argument is the fact that we have fused implementations for SDPA now. And these fused implementations allow us to easily and performantly support this new semantic. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131863
* #132710
* __->__ #131060

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by drisspg** [comment]
@justinchuby would you be able to provide some guidance for the onnx failures? Is it possible for the decomp to apply for onnx as well?

**Comment by justinchuby** [comment]
Does the decomp appear as one of the core aten decomps?

**Review by jbschlosser — COMMENTED** [review]
Nice! I'm so happy that we're on the path to finally resolving this :) Thanks for the hard work here!

At a high level, I'm still worried about the blanket use of `nan_to_num()` removing NaNs that -should- be present in the softmax output. Ideally in my head, the only NaNs that should be converted to zeros are those that arise from a fully masked-out row. This gives the safe softmax operation super clean semantics.

As is, NaNs in the input tensor will result in NaNs in the output, and these will be converted to zeros (I argue wrongly).
```python
x = torch.tensor([0.1, 0.2, 0.3, float('nan')])
output = torch.ops.aten._safe_softmax(x, dim=0)  # NaNs -> zeros without a fully masked-out row!
```

I realize that:
1. `_safe_softmax()` is private and only intended for use within attention calculations that we control
2. Inputs can be expected to come from QK^T

I don't think it's great practice to design an op where it's implicitly required to be used in a particular context. This limits the general usefulness / ability to reason about the op.

Ignoring that for now, here's an example with full SDPA for realism:

```python
import torch
from torch.nn.functional import scaled_dot_product_attention

torch.manual_seed(1)

query = torch.randn(2, 3, 4)
# a single NaN in the query input
query[0,1,2] = float('nan')
query = query.detach().requires_grad_(True)
key = torch.randn(2, 3, 4, requires_grad=True)
value = torch.randn(2, 3, 4, requires_grad=True)

# NB: no attention masking / biasing
# output -should- have NaNs; pre-this-PR, it will.
output = scaled_dot_product_attention(query, key, value, attn_mask=None)
output.sum().backward()

# query's grad -should- have NaNs; pre-this-PR, it will.
# However, the PR version of safe softmax used in SDPA will
# result in zero grads here, which arguably masks whatever issue
# caused query to have NaNs in the first place and allows training
# to continue (wrongly).
print(query.grad)
```

So my official opinion on record is that the semantics of safe softmax should be equivalent to those of softmax EXCEPT in the fully masked-out row case, where NaNs are converted to zeros. Other NaNs should be preserved, and not doing so introduces silent correctness issues that may mask issues in a way that is hard to track down / debug.

I don't want to hold this PR up by sticking too strongly to this, but I'm not positive we can claim this won't be a problem in practice, given the example above.

**Review by jbschlosser — APPROVED** [review]
I'm good with it :)

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -609,6 +610,15 @@ bool should_compute_logsumexp(const Tensor& query, const Tensor& key, const Tens
 
 } // namespace
 
+Tensor _safe_softmax(
+    const Tensor& self,
+    int64_t dim,
+    std::optional<ScalarType> dtype) {
+  auto out = at::softmax(self, dim, dtype);
+  const auto masked = self.eq(-std::numeric_limits<float>::infinity());
```
Comment (qihqi) [inline aten/src/ATen/native/transformers/attention.cpp:618]: @drisspg is it possible to put this float into a scalar tensor of dtype = torch.float32?
